### PR TITLE
Patch request

### PR DESCRIPTION
**Pull request by andyvesel**
Added patch request for mocked data

### CODE REVIEW

File: src/client/kube_store/index.ts [hunk 1]
```diff
@@ -233,6 +48,47 @@ export default class KubeStore {
     return this.instance || (this.instance = new this(''));
   }
 
+  public patchResource(resource: KubeResource, name: string, patch: object): object {
+    let source: object = {};
+    let key: string = '';
+    const myPatch: object = {'foo':'bar'};
```
Comment (eriknelson) [inline src/client/kube_store/index.ts]: Don't think we want dummy data here.

**Comment by eriknelson** [comment]
Needs rebase

File: src/client/kube_store/index.ts [hunk 2]
```diff
@@ -1,9 +1,10 @@
 import _ from 'lodash';
 import { NamespacedResource, ClusterResource, KubeResource,
   CoreNamespacedResource, CoreClusterResource, } from '../resources';
+import { ClusterClient } from '../client';
```
Comment (jwmatthews) [inline src/client/kube_store/index.ts]: I suspect this import isn't needed.

File: src/client/kube_store/index.ts (hunk 1 above)
Comment (jwmatthews) [reply src/client/kube_store/index.ts]: +1 remove this and just use 'patch' object that is passed in

File: src/client/kube_store/index.ts [hunk 3]
```diff
@@ -233,6 +48,47 @@ export default class KubeStore {
...
+        }
+      } else {
+        // Extension ClusterResource
+        key = `apis/${resource.gvk().group}/${resource.gvk().version}/${resource.gvk().kindPlural}`;
+        if (this.data()[key]) {
+          source = this.data()[key][name];
+        }
+      }
+    }
+    const patchedObject = JsonMergePatch.apply(source, myPatch);
```
Comment (jwmatthews) [inline src/client/kube_store/index.ts]: assume this should be 

const patchedObject = JsonMergePatch.apply(source, patch);

File: src/client/kube_store/index.ts [hunk 4]
```diff
@@ -233,6 +48,47 @@ export default class KubeStore {
     return this.instance || (this.instance = new this(''));
   }
 
+  public patchResource(resource: KubeResource, name: string, patch: object): object {
+    let source: object = {};
+    let key: string = '';
+    const myPatch: object = {'foo':'bar'};
+    if (resource instanceof NamespacedResource) {
```
Comment (jwmatthews) [inline src/client/kube_store/index.ts]: As we discussed, later when we know this function is working, we can explore refactoring it so that determining the key is in a different function which can be reused.

File: src/client/kube_store/index.ts [hunk 5]
```diff
@@ -233,6 +47,48 @@ export default class KubeStore {
...
+      } else {
+        // Extension ClusterResource
+        key = `apis/${resource.gvk().group}/${resource.gvk().version}/${resource.gvk().kindPlural}`;
+        if (this.data()[key]) {
+          result = this.data()[key][name];
+        }
+      }
+    }
+
+    const patchedObject = JsonMergePatch.apply(resource, patch);
```
Comment (jwmatthews) [inline src/client/kube_store/index.ts]: Patching 'resource' is incorrect.
In this context, 'resource' is an object that just tracks the GVK metadata, it does not have the actual data.

(This may be confusing as the name 'Resource' is used in k8s and typically would refer to object data, in this case we are using the class definition 'Resource' to track metadata associated with the GVK.  'Resource' in this context is essentially a class that helps us to identify the URL of the Resource on k8s API)

What you want to do is to apply the json merge patch to the data.

You want to:
1. Extract the desired data (you use the resource and it's gvk methods to form a key so you can lookup the specific data you want from localStorage).  
2. Apply the merge patch to the actual data which is in the store for this specific object

One issue in the method way it is written is that you are not getting the data.

You have a block above which is doing a lookup on this.data()[key][name], and putting the data in 'result'... yet you are not using 'result'.

I would have expected to see:
const patchedObject = JsonMergePatch.apply(result, patch);

File: src/client/kube_store/index.ts (hunk 4 above)
Comment (jwmatthews) [reply src/client/kube_store/index.ts]: Looks like you are ignoring this entire block and not using the 'result'

The 'result' is the important part of how we do the lookup of mocked data

File: src/client/kube_store/index.ts [hunk 6]
```diff
@@ -233,6 +47,48 @@ export default class KubeStore {
...
+    localStorage[key] = patchedObject;
```
Comment (jwmatthews) [inline src/client/kube_store/index.ts]: This is wrong.

As we discussed, we are storing the entire mocked data as a single key in localStorage.
So we have a single key for ALL of our data.

This method as written implies that we are storing individual elements in localStorage with a primary key, we are not doing that.  We are using a series of nested keys.

File: src/client/kube_store/index.ts [hunk 7]
```diff
@@ -233,6 +47,48 @@ export default class KubeStore {
...
+// console.log(patchedObject)
+    return resource;
```
Comment (jwmatthews) [inline src/client/kube_store/index.ts]: We wouldn't return 'resource'.

Perhaps we'd return the patchedObject or maybe we just return a success or error condition.

File: src/client/kube_store.test.ts [hunk 8]
```diff
@@ -54,3 +56,24 @@ test('Test NamespacedResource setResource', () => {
...
+  const patch = {'test':'foo'};
+
+  const expected = {
+
+      namespace: testNs,
+        _gvk:
+          { group: 'migration.openshift.io',
+            version: 'v1alpha1',
+            kindPlural: 'migclusters' },
+          test: 'foo',
```
Comment (jwmatthews) [inline src/client/kube_store.test.ts]: This doesn't line up to what we'd normally see.

We'd expect the data to be structured under a 'spec': {}
So would expect something more like
{ 'spec': {'test': 'foo'}}

File: src/client/kube_store.test.ts [hunk 9]
```diff
@@ -54,3 +56,24 @@ test('Test NamespacedResource setResource', () => {
...
+    };
+
+  const patchedObject = store.patchResource(migResource, clusterName, patch);
+  localStorage.clear();
```
Comment (jwmatthews) [inline src/client/kube_store.test.ts]: I don't follow why we'd clear localStorage

File: src/client/kube_store.test.ts [hunk 10]
```diff
@@ -54,3 +56,24 @@ test('Test NamespacedResource setResource', () => {
...
+  localStorage.setItem('test', JSON.stringify(patchedObject))
+  expect(patchedObject).toEqual(expected);
```
Comment (jwmatthews) [inline src/client/kube_store.test.ts]: We want to test how the patch command works.
So the workflow would be something like:

1) Determine some piece of data that already exists in kube_store
2) Create a patch command to modify that existing data
3) Issue the patch command
4) Check the data in kube_store and verify it was modified as you expected.

File: src/client/kube_store.test.ts (hunk 8 above)
Comment (andyvesel) [reply src/client/kube_store.test.ts]: I've added { 'spec': {patch: {'test': 'foo'}}}

File: src/client/kube_store/index.ts (hunk 7 above)
Comment (andyvesel) [reply src/client/kube_store/index.ts]: fixed

File: src/client/kube_store/index.ts (hunk 5 above)
Comment (andyvesel) [reply src/client/kube_store/index.ts]: fixed

File: src/client/kube_store/index.ts (hunk 6 above)
Comment (andyvesel) [reply src/client/kube_store/index.ts]: If I follow this correctly, the localStorage should contain the entire mocked data + patch, so it'd be like 
```
    const localStorageData = {...mocked_data, ...patchedObject};
    localStorage.setItem(localStorageMockedDataKey, JSON.stringify(localStorageData))
```

File: src/client/kube_store/index.ts [hunk 11]
```diff
@@ -221,6 +35,58 @@ export default class KubeStore {
     return d['clusters'][this.clusterName];
   }
 
+  private determineKey(resource: KubeResource, name: string): object {
```
Comment (jwmatthews) [inline src/client/kube_store/index.ts]: Note this is different than what we discussed.

We discussed that we'd refactor out a method to determine the key, hence the name of this function is 'determineKey', yet this function as written now is returning the actual value and not the key.

File: src/client/kube_store.test.ts [hunk 12]
```diff
@@ -36,6 +37,7 @@ const examplePlan = {
 const testNs = 'test-ns';
 const planName = 'my-plan';
 const expectedGvk = 'migration.openshift.io/v1alpha1/migplans';
+const clusterName = 'foo';
```
Comment (jwmatthews) [inline src/client/kube_store.test.ts]: Why is this here?

File: src/client/kube_store/index.ts [hunk 13]
```diff
@@ -233,6 +81,20 @@ export default class KubeStore {
     return this.instance || (this.instance = new this(''));
   }
 
+  public patchResource(resource: KubeResource, name: string, patch: object): object {
+    const key = this.determineKey(resource, name);
+    const currentResource = this.data()[key][name];
+    const patchedObject = JsonMergePatch.apply(currentResource, patch);
+    const localStorageData = {...mocked_data, ...patchedObject};
```
Comment (jwmatthews) [inline src/client/kube_store/index.ts]: This is not creating the structure as you want.
This is taking 'patchedObject' and putting it at the root level of the object, instead of updating it in the nested key it needs to go in.

**Comment by jwmatthews** [comment]
I will come back to help with this PR once we've gotten past the e2e milestone.

Last I looked we need to fix how patched data is being stored in localStorage.

**Comment by eriknelson** [comment]
Going to close this for now. There's probably some useful work here but it can be referenced at a future time.

**Comment by eriknelson** [comment]
Branch is available in the `patch-request` branch in `fusor/mig-ui`.
